server/repository/src/db_diesel/changelog/changelog.rs

```diff
@@ -0,0 +1,199 @@
+        name_id -> Nullable<Text>,
+        store_id -> Nullable<Text>,
+    }
+}
+
+table! {
+    changelog_deduped (cursor) {
+        cursor -> BigInt,
+        table_name -> crate::db_diesel::changelog::ChangelogTableNameMapping,
+        record_id -> Text,
+        row_action -> crate::db_diesel::changelog::ChangelogActionMapping,
+        name_id -> Nullable<Text>,
+        store_id -> Nullable<Text>,
+    }
+}
+
+#[derive(DbEnum, Debug, Clone, PartialEq, Eq)]
+#[DbValueStyle = "SCREAMING_SNAKE_CASE"]
+pub enum ChangelogAction {
+    Upsert,
+    Delete,
+}
+
+#[derive(DbEnum, Debug, Clone, PartialEq, Eq)]
+#[DbValueStyle = "snake_case"]
+pub enum ChangelogTableName {
+    Number,
+    Location,
+    StockLine,
+    Name,
+    NameStoreJoin,
+    Invoice,
+    InvoiceLine,
+    Stocktake,
+    StocktakeLine,
+    Requisition,
+    RequisitionLine,
+}
+
+#[derive(Clone, Queryable, Debug, PartialEq, Insertable)]
+#[table_name = "changelog"]
+pub struct ChangelogRow {
+    pub cursor: i64,
+    pub table_name: ChangelogTableName,
+    pub record_id: String,
+    pub row_action: ChangelogAction,
+    pub name_id: Option<String>,
+    pub store_id: Option<String>,
+}
+
+#[derive(Default, Clone)]
+pub struct ChangelogFilter {
+    pub table_name: Option<EqualFilter<ChangelogTableName>>,
+    pub name_id: Option<EqualFilter<String>>,
+    pub store_id: Option<EqualFilter<String>>,
+    pub record_id: Option<EqualFilter<String>>,
+}
+
+pub struct ChangelogRepository<'a> {
+    connection: &'a StorageConnection,
+}
+
+impl<'a> ChangelogRepository<'a> {
+    pub fn new(connection: &'a StorageConnection) -> Self {
+        ChangelogRepository { connection }
+    }
+
+    /// Returns changelog rows order by operation sequence in asc order
+    ///
+    /// # Arguments
+    ///
+    /// * `earliest` - Starting cursor (first returned changelogs may be ahead in sequence from starting cursor)
+    /// * `limit` - Maximum number of entries to be returned
+    /// * `filter` - Extra filter to apply on change_logs
+    pub fn changelogs(
+        &self,
+        earliest: u64,
+        limit: u32,
+        filter: Option<ChangelogFilter>,
+    ) -> Result<Vec<ChangelogRow>, RepositoryError> {
+        let query = create_filtered_query(earliest, filter).limit(limit.into());
+
+        // // Debug diesel query
+        // println!(
+        //     "{}",
+        //     diesel::debug_query::<crate::DBType, _>(&query).to_string()
+        // );
+
+        let result = query.load(&self.connection.connection)?;
+        Ok(result)
+    }
+
+    pub fn count(
+        &self,
+        earliest: u64,
+        filter: Option<ChangelogFilter>,
+    ) -> Result<u64, RepositoryError> {
+        let result = create_filtered_query(earliest, filter)
+            .count()
+            .get_result::<i64>(&self.connection.connection)?;
+        Ok(result as u64)
+    }
+
+    // TODO we only really want to know the cursor, so just max ?
```

Comment:
changed to `max` rather then last change log, also comment removed here: c9817e73fbae99e50ec101bb9daa4f8c6fbad177

Reply:
I'd be tempted to rename to something like `latest_cursor`?

I'd also found myself looking up what the function is used for because I assumed `normally` we wouldn't get the max cursor, I wonder if it's worth a quick comment to explain what this is used for e.g. 
`// After initial sync we use this method to get the latest cursor to make sure we don't try to re-process any records that were synced to this site`

Definitely not important but I was confused why this existed :)

Reply:
Good idea, thank you 0367bb6b0aad24f3f395bb91c1bf76c33ba240c1

> Definitely not important but I was confused why this existed :)

Keen to get feedback about confusing bits, for code author it's so hard to see those.